apidef/oas/upstream.go

```diff
@@ -35,12 +36,19 @@ func (u *Upstream) Fill(api apidef.APIDefinition) {
 		u.Test = nil
 	}
 
-	u.Certificates = make([]Certificate, len(api.UpstreamCertificates))
+	u.Certificates = make(Certificates, len(api.UpstreamCertificates))
 	u.Certificates.Fill(api.UpstreamCertificates)
 
 	if len(u.Certificates) == 0 {
 		u.Certificates = nil
 	}
+
+	u.PinnedPublicKeys = make(PinnedPublicKeys, len(api.PinnedPublicKeys))
+	u.PinnedPublicKeys.Fill(api.PinnedPublicKeys)
+
+	if len(u.PinnedPublicKeys) == 0 {
+		u.PinnedPublicKeys = nil
```

Comment:
Could we just avoid this whole if case by checking `if len(api.PinnerPublicKeys) > 0` around it? It would avoid both the allocation and the dereference here

Reply:
Same goes for api.UpstreamCertificates just above this

Reply:
That's correct, let me change that.

Reply:
Well, there is a case where  `len(api.PinnerPublicKeys) = 0` but `u.PinnedPublicKeys` is not empty. In that case, we should set `u.PinnedPublicKeys` `nil` otherwise update operations will not work.

Reply:
So, the proposal will not handle that scenario